### PR TITLE
Current level gauge for FlyteWorkflows

### PR DESCRIPTION
**Pull request by wild-endeavor**
# TL;DR
First in a long series of work around Propeller observability and system observability as a whole.  Refer to the GH issue for more information.

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [x] Unit tests added
 - [x] Code documentation added
 - [x] Any pending items have an associated Issue

## Complete description
* Create a new class that keeps a lister of FlyteWorkflow objects and periodically emits metrics about the list.
* Also changing the port for local development for Minio to reflect the new sandbox `NodePort`.

## Tracking Issue
https://github.com/lyft/flyte/issues/311

## Follow-up issue
NA

### CODE REVIEW

File: pkg/controller/controller.go [hunk 1]
```diff
@@ -169,6 +179,97 @@ func (c *Controller) getWorkflowUpdatesHandler() cache.ResourceEventHandler {
...
+	// Emit gauges at both the project/domain level, but also at the aggregate level
+	workflows, err := r.lister.List(labels.Everything())
+	if err != nil {
+		logger.Errorf(ctx, "Error listing workflows when attempting to collect data for gauges %s", err)
+	}
+
+	counts := r.countList(ctx, workflows)
+
+	// Emit labeled metrics, for each project/domain combination. This can be aggregated later with Prometheus queries.
+	metricKeys := []contextutils.Key{contextutils.ProjectKey, contextutils.DomainKey}
```
Comment (wild-endeavor) [inline pkg/controller/controller.go:229]: Decided not to do phase for now @EngHabu - should we move this to stdlib?  It's not quite the same in the sense that I don't really care to emit the _unlabeled metric.

File: pkg/controller/controller.go [hunk 2]
```diff
@@ -169,6 +179,99 @@ func (c *Controller) getWorkflowUpdatesHandler() cache.ResourceEventHandler {
 	}
 }
 
+// This object is responsible for emitting metrics that show the current number of Flyte workflows, cut by project and domain.
+// It contains
```
Comment (EngHabu) [inline pkg/controller/controller.go:183]: typo

File: pkg/controller/controller.go [hunk 3]
```diff
@@ -169,6 +179,99 @@ func (c *Controller) getWorkflowUpdatesHandler() cache.ResourceEventHandler {
...
+type ResourceLevelMonitor struct {
+	Scope promutils.Scope
+
+	// Meta timer - this times each collection cycle
+	CollectorTimer promutils.StopWatch
+
+	// This is a labeled gauge
+	levels *prometheus.GaugeVec
+
+	// The thing that we want to current levels of
```
Comment (EngHabu) [inline pkg/controller/controller.go:194]: nit: ?
```suggestion
	// The thing that we want to measure the current levels of
```

File: pkg/controller/controller.go [hunk 4]
```diff
@@ -169,6 +179,99 @@ func (c *Controller) getWorkflowUpdatesHandler() cache.ResourceEventHandler {
...
+
+func (r *ResourceLevelMonitor) countList(ctx context.Context, workflows []*v1alpha1.FlyteWorkflow) map[string]map[string]int {
+	// Map of Projects to Domains to counts
+	counts := map[string]map[string]int{}
+
+	// Collect all workflow metrics
+	for i, wf := range workflows {
+		execID := wf.GetExecutionID()
+		if execID.WorkflowExecutionIdentifier == nil {
+			logger.Warningf(ctx, "Workflow does not have an execution identifier! [%v]", wf)
```
Comment (EngHabu) [inline pkg/controller/controller.go:206]: Should we still count these under "unknown-invalid" project/domain? I expect this to be 0... but would be curious to know if it's not

File: pkg/controller/controller.go (hunk 1 above)
Comment (EngHabu) [reply pkg/controller/controller.go:231]: unlabeled is an option.. when you create any of the labeled.New*, if you do not specify EmitUnlabeledOption, it won't...

File: pkg/controller/controller.go [hunk 5]
```diff
@@ -169,6 +179,99 @@ func (c *Controller) getWorkflowUpdatesHandler() cache.ResourceEventHandler {
...
+	}
+
+	counts := r.countList(ctx, workflows)
+
+	// Emit labeled metrics, for each project/domain combination. This can be aggregated later with Prometheus queries.
+	metricKeys := []contextutils.Key{contextutils.ProjectKey, contextutils.DomainKey}
+	for project, val := range counts {
+		for domain, num := range val {
+			tempContext := contextutils.WithProjectDomain(ctx, project, domain)
+			gauge, err := r.levels.GetMetricWith(contextutils.Values(tempContext, metricKeys...))
```
Comment (EngHabu) [inline pkg/controller/controller.go:235]: I presume you verified/test this works? and won't just keep a single value (the latest) around?

File: pkg/controller/controller.go [hunk 6]
```diff
@@ -169,6 +179,99 @@ func (c *Controller) getWorkflowUpdatesHandler() cache.ResourceEventHandler {
...
+			if err != nil {
+				panic(err)
+			}
+			gauge.Set(float64(num))
+		}
+	}
+}
+
+func (r *ResourceLevelMonitor) RunCollector(ctx context.Context) {
+	ticker := time.NewTicker(5 * time.Second)
```
Comment (EngHabu) [inline pkg/controller/controller.go:245]: maybe move it to a const up top?

File: pkg/controller/controller.go [hunk 7]
```diff
@@ -169,6 +179,99 @@ func (c *Controller) getWorkflowUpdatesHandler() cache.ResourceEventHandler {
...
+	collectorCtx := contextutils.WithGoroutineLabel(ctx, "resource-level-monitor")
+	pprof.SetGoroutineLabels(collectorCtx)
```
Comment (EngHabu) [inline pkg/controller/controller.go:247]: Do we need this line?

File: pkg/controller/controller.go [hunk 8]
```diff
@@ -169,6 +179,99 @@ func (c *Controller) getWorkflowUpdatesHandler() cache.ResourceEventHandler {
...
+				t := r.CollectorTimer.Start()
+				r.collect(collectorCtx)
+				t.Stop()
+			}
+		}
+	}()
+}
+
+func NewResourceLevelMonitor(scope promutils.Scope, lister lister.FlyteWorkflowLister) *ResourceLevelMonitor {
+	gauge := scope.MustNewGaugeVec("flyteworkflow", "Current FlyteWorkflow levels", contextutils.ProjectKey.String(),
```
Comment (EngHabu) [inline pkg/controller/controller.go:264]: nit: move it to the object initialization below?

File: pkg/controller/controller.go (hunk 7 above)
Comment (wild-endeavor) [reply pkg/controller/controller.go:245]: i have no idea what this line does, i just copied it from the worker goroutines.  Should i delete?

File: pkg/controller/controller.go (hunk 5 above)
Comment (wild-endeavor) [reply pkg/controller/controller.go:233]:
```
# HELP flyte:propeller:all:collector:collection_cycle_ms Measures how long it takes to run a collection
# TYPE flyte:propeller:all:collector:collection_cycle_ms summary
flyte:propeller:all:collector:collection_cycle_ms{quantile="0.5"} 0
flyte:propeller:all:collector:collection_cycle_ms{quantile="0.9"} 0
flyte:propeller:all:collector:collection_cycle_ms{quantile="0.99"} 0
flyte:propeller:all:collector:collection_cycle_ms_sum 0
flyte:propeller:all:collector:collection_cycle_ms_count 9
# HELP flyte:propeller:all:collector:flyteworkflow Current FlyteWorkflow levels
# TYPE flyte:propeller:all:collector:flyteworkflow gauge
flyte:propeller:all:collector:flyteworkflow{domain="development",project="flyteexamples"} 1
flyte:propeller:all:collector:flyteworkflow{domain="staging",project="flyteexamples"} 1
```

File: pkg/controller/controller.go (hunk 4 above)
Comment (wild-endeavor) [reply pkg/controller/controller.go:211]: added as 'missing'
